cpp/src/sort.cpp

```diff
@@ -93,10 +95,234 @@ std::unique_ptr<cudf::column> get_split_ind(TaskContext& ctx,
+  TaskContext& ctx, std::shared_ptr<arrow::Table> global_split_candidates)
+{
+  auto split_indices = vector_to_array(
+    get_split_ind(ctx, global_split_candidates->num_rows(), ctx.nranks, /* include_start */ false));
+  auto split_values =
+    ARROW_RESULT(arrow::compute::Take(global_split_candidates, split_indices)).table();
+  return std::move(split_values);
+}
+
+// This function is poorly optimised but arrow doesn't give us other options
+// Compare two tables each containing a single row
+bool compare(TaskContext& ctx,
+             std::shared_ptr<arrow::Table> row_a,
+             std::shared_ptr<arrow::Table> row_b,
+             const std::vector<cudf::size_type>& keys_idx,
+             const std::vector<arrow::compute::SortKey>& column_order,
+             arrow::compute::NullPlacement null_precedence)
+{
+  // Create a table with each row, sort that table
+  // Use the resulting order as the comparison
+  // We are testing row_a < row_b
+  // So if row_a gets sorted into position 0, then row_a < row_b
+
+  auto combined = ARROW_RESULT(arrow::ConcatenateTables({row_b, row_a}));
+
+  auto sort_indices = ARROW_RESULT(arrow::compute::SortIndices(
+    combined, arrow::compute::SortOptions(column_order, null_precedence)));
+
+  std::shared_ptr<arrow::UInt64Scalar> first_sort_index =
+    std::dynamic_pointer_cast<arrow::UInt64Scalar>(ARROW_RESULT(sort_indices->GetScalar(0)));
+  return first_sort_index->value == 1;
+}
+
+std::size_t lower_bound_row(TaskContext& ctx,
+                            std::shared_ptr<arrow::Table> haystack,
+                            std::shared_ptr<arrow::Table> needle,  // This is a single row
+                            const std::vector<cudf::size_type>& keys_idx,
+                            const std::vector<arrow::compute::SortKey>& column_order,
+                            arrow::compute::NullPlacement null_precedence)
+{
+  std::size_t first  = 0;
+  std::size_t length = haystack->num_rows();
+  while (length > 0) {
+    auto rem = length % 2;
+    length /= 2;
+    if (compare(ctx,
+                haystack->Slice(first + length, 1),
+                needle,
+                keys_idx,
+                column_order,
+                null_precedence)) {
+      first += length + rem;
+    }
+  }
+  return first;
+}
+
+std::vector<std::size_t> find_destination_ranks(
+  TaskContext& ctx,
+  std::shared_ptr<arrow::Table> sorted_table,
+  std::shared_ptr<arrow::Table> global_split_values,
+  const std::vector<cudf::size_type>& keys_idx,
+  const std::vector<arrow::compute::SortKey>& column_order,
+  arrow::compute::NullPlacement null_precedence)
+{
+  // Create a new table with the rank column appended
+  auto rank_array = create_array<int64_t>(sorted_table->num_rows(), ctx.rank);
+  auto sorted_table_with_rank =
+    ARROW_RESULT(sorted_table->AddColumn(sorted_table->num_columns(),
+                                         arrow::field("split_rank", rank_array->type()),
```

Comment:
I guess we need a name, although, in theory it could collide with an existing one.  Could see taking on this as a small debt.  Making a new name isn't that trivial I guess (unless we just make it longer), since you would have to e.g. use `("_" * max_len) + "split_rank"` or so.